lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,316 @@
+//===-- lib/semantics/check-select-stmt.cc --------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "check-select-stmt.h"
+#include "flang/parser/message.h"
+#include "flang/semantics/tools.h"
+
+namespace Fortran::semantics {
+
+class SelectCaseHelper {
+
+public:
+  SelectCaseHelper(parser::Messages &messages) : messages_{messages} {}
+
+  void ReportOverlapCaseValError(
+      parser::CharBlock &source1, parser::CharBlock &source2) {
+    // C1149
+    messages_
+        .Say(source2,
+            "SELECT CASE statement value at this location overlaps with the below location"_err_en_US)
+        .Attach(source1,
+            "SELECT CASE statement value overlaps with the above location"_err_en_US);
+  }
+
+  void Insert(std::int64_t intVal, parser::CharBlock source) {
+    auto selectCaseIntRangeValsIter =
+        std::find_if(selectCaseIntRangeVals_.begin(),
+            selectCaseIntRangeVals_.end(), [&intVal](const auto &mem) {
+              return ((!mem.fromVal_ || (intVal >= mem.fromVal_.value())) &&
+                  (!mem.toVal_ || (intVal <= mem.toVal_.value())));
+            });
+    if (selectCaseIntRangeValsIter != selectCaseIntRangeVals_.end()) {
+      ReportOverlapCaseValError(selectCaseIntRangeValsIter->source_, source);
+    }
+    auto selectCaseIntValsInsertStatus =
+        selectCaseIntVals_.insert(std::make_pair(intVal, source));
+    if (selectCaseIntValsInsertStatus.second == false) {
```

Comment:
The logical negation operator `!` would be better idiomatic C/C++ usage here rather than an equality test with `false`.

Reply:
I have done this changes.